src/main/java/org/junitpioneer/jupiter/ResourceManagerExtension.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2016-2021 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.platform.commons.support.ReflectionSupport;
+
+final class ResourceManagerExtension implements ParameterResolver {
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		// TODO: Also return true if the parameter is annotated with @Shared
+		// TODO: Check that the parameter is not annotated with both @New and @Shared.
+		return parameterContext.isAnnotated(New.class);
+	}
+
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		// TODO: Check that the parameter is annotated with at least @New or @Shared.
+		// TODO: Check that the parameter is not annotated with both @New and @Shared.
+		New newResourceAnnotation = parameterContext.findAnnotation(New.class).orElseThrow(() -> {
+			throw new ParameterResolutionException(String
+					.format( //
+						"Parameter `%s` on %s is not annotated with @New", //
+						parameterContext.getParameter(), //
+						extensionContext
+								.getTestMethod()
+								.map(method -> "method `" + method + '`')
+								.orElse("unknown method")));
+		});
+		ResourceFactory<?> resourceFactory =
+				ReflectionSupport.newInstance(newResourceAnnotation.value(), (Object[]) newResourceAnnotation.arguments());
+		extensionContext
+				.getStore(
+					ExtensionContext.Namespace.create(ResourceManagerExtension.class, extensionContext.getUniqueId()))
+				.put(System.identityHashCode(newResourceAnnotation) + "-resource-factory",
+					(ExtensionContext.Store.CloseableResource) resourceFactory::close);
+		Resource<?> resource;
+		try {
+			resource = resourceFactory.create();
+		}
+		catch (Exception e) {
+			throw new ParameterResolutionException(
+				"Unable to create an instance of `" + resourceFactory.getClass() + '`', e);
+		}
+		extensionContext
+				.getStore(
+					ExtensionContext.Namespace.create(ResourceManagerExtension.class, extensionContext.getUniqueId()))
+				.put(System.identityHashCode(newResourceAnnotation) + "-resource",
```

Comment:
As above, but one more option:

* ask the resource for a unique name - by default it could return its own `identityHashCode` (or the  random long), but certain resources may already be unique (e.g. file path for `Resource<Path>`)

Reply:
I think using a random `long`/`AtomicLong` is easiest to understand, so unless you have any objections, then that's the method I will try first. :)